MAVProxy/modules/mavproxy_DGPS.py

```diff
@@ -1,24 +1,119 @@
+                                     self.dgps_settings.completion)
+
+        self.port = None      
+        self.last_pkt = None 
+        self.inject_seq_nr = 0
+
+    def log_rtcm(self, data):
+        '''optionally log rtcm data'''
+        if self.dgps_settings.logfile is None:
+            return
+        if self.logfile is None:
+            self.logfile = open(self.dgps_settings.logfile, 'wb')
+        if self.logfile is not None:
+            self.logfile.write(data)
+
+    def cmd_dgps(self, args):
+        '''dgps command handling'''
+        if len(args) <= 0:
+            print("Usage: dgps <start|stop|status|set>")
+            return
+        if args[0] == "start":
+            self.cmd_start()
+        if args[0] == "stop":
+            self.port = None
+        elif args[0] == "status":
+            self.dgps_status()
+        elif args[0] == "set":
+            self.dgps_settings.command(args[1:])
+
+    def cmd_start(self):
+        '''start dgps link'''
+        if self.dgps_settings.conntype == "UDP":
+            print("Connecting to UDP RTCM Stream")
+            self.connect_udp_rtcm_base(self.dgps_settings.ip, self.dgps_settings.port)
+        elif self.dgps_settings.conntype == "TCP":
+            print("Connecting to TCP RTCM Stream")
+            self.connect_tcp_rtcm_base(self.dgps_settings.ip, self.dgps_settings.port)
+        else:
+            print("Undefined connection type!")
+            return           
+
+    def dgps_status(self):
+        '''show dgps status'''
+        now = time.time()
+        if self.port is None:
+            print("DGPS: Not started")
+            return
+        elif self.last_pkt is None:
+            print("DGPS: no data")
+            return
+        
+        print ("Last packet recieved %.3fs ago" % (now - self.last_pkt))
+        # frame_size = 0
+        # for id in sorted(self.id_counts.keys()):
+        #     print(" %4u: %u (len %u)" % (id, self.id_counts[id], len(self.last_by_id[id])))
+        #     frame_size += len(self.last_by_id[id])
+        # print("dgps: %u packets, %.1f bytes/sec last %.1fs ago framesize %u" % (self.pkt_count, self.rate, now - self.last_pkt, frame_size))
+
+    def connect_udp_rtcm_base(self, ip, port):
+        print "UDP Connection"
         self.portnum = 13320
         self.port = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
         self.port.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
         self.port.bind(("127.0.0.1", self.portnum))
         mavutil.set_close_on_exec(self.port.fileno())
         self.port.setblocking(0)
-        self.inject_seq_nr = 0
         print("DGPS: Listening for RTCM packets on UDP://%s:%s" % ("127.0.0.1", self.portnum))
-    
+
+    def connect_tcp_rtcm_base(self, ip, port):
+        print "TCP Connection"
+        try:
+            self.port = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            self.port.connect((ip, port))
+            self.port.settimeout(1)
```

Comment:
need to make this non-blocking

Reply:
I made the change and the lost packet issue did go away. Thanks for the suggestion Tridge. PR is updated.